torch/fx/experimental/symbolic_shapes.py

```diff
@@ -916,9 +942,258 @@ def _print_Symbol(self, expr) -> str:
+        return f"{info.name}{index_expr}"
+
+    def compile(self, exprs: List[str], python_exprs: List[str]) -> "GuardCompiler.SymbolicShapesExpr":
+        """Generates and compiles the guard function in C++.
+
+        1. Generates a C++ function based on its object's registered parameters
+        2. Compiles the C++ function, and loads it with the help of `CppCodeCache`
+
+        The overall structure of the generated function is as follows:
+
+        extern "C" int guard(PyObject** inputs) {
+            // Inputs unpacking
+            // Source creation (e.g. sizes, strides, and storage offset variables are defined)
+            // Return the actual guard expression
+        }
+        """
+        from torch._dynamo.source import TensorPropertySource, TensorProperty
+
+        if len(exprs) == 0:
+            return self.SymbolicShapesExpr([], "True")
+
+        fn_name = "___symbolic_shape_fn"
+
+        # 1. Function call
+        #     a) Get all the argument expressions for each parameter
+        call_arguments = ", ".join(info.source.name() for info in self.parameter_infos)
+        #     b) Generate the actual function call
+        call_expression = f"{fn_name}({call_arguments})"
+
+        # 2. Function definition
+        #     a) Generate the actual guard expression
+        guard_expression = " && ".join(exprs)
+        #     b) Generate the extra temporary variables corresponding to Tensor
+        #        properties (e.g. sizes, strides, and storage_offset)
+        source_creation = []
+        for i, info in enumerate(self.parameter_infos):
+            if info.type == SymInt:
+                source_creation.append(f"int64_t {info.name} = THPUtils_unpackLong(inputs[{i}]);")
+
+            elif info.type == torch.Tensor:
+                tensor_name = self.register_unique_name(f"{info.name}_tensor")
+                source_creation.append(f"const auto& {tensor_name} = THPVariable_Unpack(inputs[{i}]);")
+
+                for tensor_property in TensorProperty:
+                    if tensor_property == TensorProperty.STORAGE_OFFSET:
+                        tensor_property_str = "storage_offset"
+                        tensor_property_source = TensorPropertySource(info.source, tensor_property)
+                    else:
+                        tensor_property_str = f"{tensor_property.name.lower()}s"
+                        tensor_property_source = TensorPropertySource(info.source, tensor_property, idx=0)
+
+                    if not self.is_source_registered(tensor_property_source):
+                        # Skip if there are no SymInts tracked for this property.
+                        # Usually true for scalar tensors.
+                        continue
+
+                    tensor_property_info = self.get_source_info(tensor_property_source)
+                    source_creation.append(
+                        f"auto {tensor_property_info.name} = {tensor_name}.{tensor_property_str}();"
+                    )
```

Comment:
If the 4 LOC per tensor is a concern, can you not cut that to 1 LOC using a helper:
```cpp
struct TensorProperties {
    IntArrayRef sizes;
    IntArrayRef strides;
    int64_t storage_offset;
};
TensorProperties get_tensor_properties(PyObject* obj) {
    const Tensor& tensor = THPVariable_Unpack(obj);
    return TensorProperties{
        tensor.sizes(),
        tensor.strides(),
        tensor.storage_offset()
    };
}
```

Even more extreme, you could split `inputs` into two arrays: one for ints and one for tensors. Then you can unpack into arrays with a constant number of lines:
```cpp
int guard(PyObject** int_inputs, PyObject** tensor_inputs) {
    int64_t int_params[10];
    for (int64_t i = 0; i < 10; ++i) {
        int_params[i] = THPVariable_unpackLong(int_inputs[i]);
    }
    TensorProperties tensor_params[564];
    for (int64_t i = 0; i < 564; ++i) {
       tensor_params[i] = get_tensor_properties(tensor_inputs[i]);
    }
    return ...;
}
```

Reply:
I did try this out, but it seemed to make compilation slower. My guess is that the compiler is unrolling the static loop => slower compile times.